File: src/components/widgets/skinRollerGeometry.ts
/**
 * The physics and layout maths behind the widget skin roller, kept free of
 * React and the DOM so the drum's behavior can be tested directly.
 *
 * One number drives everything: `offset`, the distance the drum has been
 * rolled, in pixels. `offset / ROW_HEIGHT` is the (fractional) index sitting
 * in the selection lane, so offset 0 is the first skin and the drum always
 * settles on a whole multiple of ROW_HEIGHT.
 */

/** Height of one skin row, and therefore the travel between two detents. */
export const ROW_HEIGHT = 52

/** Degrees of barrel rotation between neighbouring rows. */
export const ANGLE_STEP = 20

/**
 * Barrel radius that makes ANGLE_STEP of rotation advance the surface by
 * exactly one row, so the curved drum and the flat fallback move in step.
 */
export const DRUM_RADIUS = ROW_HEIGHT / (2 * Math.tan((ANGLE_STEP * Math.PI) / 360))

/** Rows further than this from the lane have rotated past the horizon. */
const HORIZON = 90 / ANGLE_STEP

/** Resistance constant for the stretch past either end. Lower drags harder. */
const RUBBER = 0.55

/**
 * How far a drag past the end is actually allowed to move the drum: the pull
 * is fed through a curve that approaches `limit` but never reaches it, so the
 * end always feels like a wall with give rather than a wall.
 */
export function rubberBand(overshoot: number, limit = ROW_HEIGHT): number {
  if (overshoot <= 0) return 0
  return (1 - 1 / ((overshoot / limit) * RUBBER + 1)) * limit
}

/** Applies the end resistance to a raw dragged offset. */
export function resistOffset(rawOffset: number, count: number, rowHeight = ROW_HEIGHT): number {
  const max = Math.max(0, count - 1) * rowHeight
  if (rawOffset < 0) return -rubberBand(-rawOffset, rowHeight)
  if (rawOffset > max) return max + rubberBand(rawOffset - max, rowHeight)
  return rawOffset
}

/** True while the drum is stretched past either end. */
export function isPastEnd(offset: number, count: number, rowHeight = ROW_HEIGHT): boolean {
  return offset < 0 || offset > Math.max(0, count - 1) * rowHeight
}

/** Which row is in the lane right now — always a real, in-range index. */
export function indexForOffset(offset: number, count: number, rowHeight = ROW_HEIGHT): number {
  if (count <= 0) return 0
  return Math.min(count - 1, Math.max(0, Math.round(offset / rowHeight)))
}

/** Where the drum comes to rest from here: the nearest whole row. */
export function settledOffset(offset: number, count: number, rowHeight = ROW_HEIGHT): number {
  return indexForOffset(offset, count, rowHeight) * rowHeight
}

/** Rows this far from the lane or nearer stay perfectly sharp. */
const SHARP_ROWS = 3

/** The full out-of-focus blur, in pixels, worn by every row past the ramp. */
const EDGE_BLUR = 6

/** How many rows past the sharp zone it takes to reach the full blur. */
const BLUR_RAMP_ROWS = 1

export interface RowPlacement {
  /** Barrel rotation for this row, in degrees. 0 means it is in the lane. */
  rotateX: number
  /** Flat-list fallback travel, in pixels, used when motion is reduced. */
  translateY: number
  opacity: number
  /** Softening at the drum's far edges, so it dissolves instead of ending. */
  blur: number
  /** Rows nearer the lane paint over rows behind them. */
  zIndex: number
  /** Rotated past the horizon — skip it entirely rather than paint a sliver. */
  hidden: boolean
}

/** Places one row relative to the current offset. */
export function placeRow(index: number, offset: number, rowHeight = ROW_HEIGHT): RowPlacement {
  const distance = index - offset / rowHeight
  const away = Math.abs(distance)
  return {
    // `|| 0` only normalises the -0 the lane row would otherwise produce.
    rotateX: -distance * ANGLE_STEP || 0,
    translateY: distance * rowHeight,
    // Fades to nothing by the time a row reaches the horizon, so rows leave
    // the drum rather than blinking out.
    opacity: away >= HORIZON ? 0 : Math.max(0, 1 - (away / HORIZON) ** 1.4),
    // The lane and three choices each way are sharp; just past the third the
    // blur ramps up over one row and then holds flat, so everything further
    // out is plainly out of focus rather than merely hinted at.
    blur: Math.min(EDGE_BLUR, Math.max(0, (away - SHARP_ROWS) / BLUR_RAMP_ROWS) * EDGE_BLUR),
    zIndex: Math.max(0, 100 - Math.round(away * 10)),
    hidden: away >= HORIZON,
  }
}

/** Wheel travel that counts as one deliberate notch of scrolling. */
export const WHEEL_NOTCH = 40

/**
 * Wheel scrolling moves the drum one whole row at a time rather than dragging
 * it continuously: accumulated wheel travel is converted into whole steps, and
 * whatever is left over stays banked for the next event. A trackpad flick and
 * a mouse notch therefore both advance in clean, animated single rows.
 */
export function wheelSteps(accumulated: number, notch = WHEEL_NOTCH): { steps: number; remainder: number } {
  const steps = Math.trunc(accumulated / notch)
  return { steps, remainder: accumulated - steps * notch }
}

/**
 * One frame of the settle spring: a critically damped pull toward `target`
 * that never overshoots enough to cross a detent it did not earn.
 */
export function stepSettle(offset: number, target: number, elapsedMs: number): number {
  const remaining = target - offset
  if (Math.abs(remaining) < 0.5) return target
  // Frame-rate independent exponential approach (~150 ms to arrive).
  const t = 1 - Math.exp(-elapsedMs / 55)
  return offset + remaining * t
}

File: apple/Sources/GrovepadChrome/Surfaces/SkinRollerGeometry.swift
import Foundation

// ---------------------------------------------------------------------------
// The physics and layout maths behind the skin roller: a line-for-line port
// of `skinRollerGeometry.ts`, kept free of SwiftUI so the drum can be tested
// directly.
//
// One number drives everything: `offset`, the distance the drum has been
// rolled, in points. `offset / rowHeight` is the (fractional) index sitting
// in the selection lane, so offset 0 is the first skin and the drum always
// settles on a whole multiple of `rowHeight`.
//
// The one native addition is `project`: the web hands `rotateX` +
// `translateZ` to CSS 3D; SwiftUI has no shared perspective between sibling
// views, so the same barrel is projected here by hand (a row's lift, its
// magnification and its foreshortening) with the same radius and perspective.
// ---------------------------------------------------------------------------

public enum SkinRollerGeometry {
    /// Height of one skin row, and therefore the travel between two detents.
    public static let rowHeight: Double = 52

    /// Degrees of barrel rotation between neighbouring rows (the web's
    /// legibility budget: at 12° the third row out keeps four fifths of its
    /// height once projected).
    public static let angleStep: Double = 12

    /// Barrel radius that makes `angleStep` advance the surface by one row.
    public static let drumRadius: Double = rowHeight / (2 * tan(angleStep * .pi / 360))

    /// How much the projection magnifies the lane row.
    public static let projectionScale: Double = 1.312

    /// The perspective distance that projects this barrel at `projectionScale`.
    public static let drumPerspective: Double = (drumRadius / (1 - 1 / projectionScale)).rounded()

    /// Rows further than this from the lane have rotated past the horizon.
    static let horizon: Double = 90 / angleStep

    /// How far from the lane the drum dissolves (inside the horizon).
    static let fadeRows: Double = min(6, horizon)

    /// Resistance constant for the stretch past either end. Lower drags harder.
    static let rubber: Double = 0.55

    /// A pull past the end, fed through a curve that approaches `limit` but
    /// never reaches it: a wall with give.
    public static func rubberBand(_ overshoot: Double, limit: Double = rowHeight) -> Double {
        guard overshoot > 0 else { return 0 }
        return (1 - 1 / ((overshoot / limit) * rubber + 1)) * limit
    }

    /// Applies the end resistance to a raw dragged offset.
    public static func resistOffset(_ raw: Double, count: Int, rowHeight: Double = rowHeight) -> Double {
        let max = Double(Swift.max(0, count - 1)) * rowHeight
        if raw < 0 { return -rubberBand(-raw, limit: rowHeight) }
        if raw > max { return max + rubberBand(raw - max, limit: rowHeight) }
        return raw
    }

    /// True while the drum is stretched past either end.
    public static func isPastEnd(_ offset: Double, count: Int, rowHeight: Double = rowHeight) -> Bool {
        offset < 0 || offset > Double(max(0, count - 1)) * rowHeight
    }

    /// Which row is in the lane right now — always a real, in-range index.
    public static func index(forOffset offset: Double, count: Int, rowHeight: Double = rowHeight) -> Int {
        guard count > 0 else { return 0 }
        // JavaScript's Math.round: halves round up, including negative ones.
        let rounded = Int((offset / rowHeight + 0.5).rounded(.down))
        return min(count - 1, max(0, rounded))
    }

    /// Where the drum comes to rest from here: the nearest whole row.
    public static func settledOffset(_ offset: Double, count: Int, rowHeight: Double = rowHeight) -> Double {
        Double(index(forOffset: offset, count: count, rowHeight: rowHeight)) * rowHeight
    }

    /// Rows this far from the lane or nearer stay perfectly sharp.
    static let sharpRows: Double = 3
    /// The full out-of-focus blur worn by every row past the ramp.
    static let edgeBlur: Double = 6
    /// How many rows past the sharp zone it takes to reach the full blur.
    static let blurRampRows: Double = 1

    public struct RowPlacement: Equatable, Sendable {
        /// Barrel rotation, in degrees. 0 means the row is in the lane.
        public var rotateX: Double
        /// Flat-list travel, used when motion is reduced.
        public var translateY: Double
        public var opacity: Double
        public var blur: Double
        public var zIndex: Double
        public var hidden: Bool
    }

    /// `placeRow`: one row relative to the current offset.
    public static func placeRow(_ index: Int, offset: Double, rowHeight: Double = rowHeight) -> RowPlacement {
        let distance = Double(index) - offset / rowHeight
        let away = abs(distance)
        let rotate = -distance * angleStep
        return RowPlacement(
            rotateX: rotate == 0 ? 0 : rotate,
            translateY: distance * rowHeight,
            opacity: away >= fadeRows ? 0 : max(0, 1 - pow(away / fadeRows, 1.4)),
            blur: min(edgeBlur, max(0, (away - sharpRows) / blurRampRows) * edgeBlur),
            zIndex: max(0, 100 - (away * 10).rounded()),
            hidden: away >= fadeRows
        )
    }

    /// Where a placed row lands on screen, relative to the lane's centre line.
    public struct Projection: Equatable, Sendable {
        /// Vertical travel of the row's centre from the lane's centre.
        public var y: Double
        /// Magnification from riding nearer the eye (the lane's is `projectionScale`).
        public var scale: Double
        /// Foreshortening: how much of its height a tilted row keeps.
        public var squash: Double
    }

    /// The barrel projected by hand: a row rotated `rotateX` degrees and pushed
    /// out `drumRadius`, seen from `drumPerspective` away.
    public static func project(_ placement: RowPlacement, reducedMotion: Bool = false) -> Projection {
        if reducedMotion { return Projection(y: placement.translateY, scale: 1, squash: 1) }
        let angle = -placement.rotateX * .pi / 180
        let depth = drumRadius * cos(angle)
        let scale = drumPerspective / (drumPerspective - depth)
        return Projection(y: drumRadius * sin(angle) * scale, scale: scale, squash: max(0, cos(angle)))
    }

    /// Wheel travel that counts as one deliberate notch of scrolling.
    public static let wheelNotch: Double = 40

    /// Accumulated wheel travel converted into whole row steps; the rest stays banked.
    public static func wheelSteps(_ accumulated: Double, notch: Double = wheelNotch) -> (steps: Int, remainder: Double) {
        let steps = (accumulated / notch).rounded(.towardZero)
        return (Int(steps), accumulated - steps * notch)
    }

    /// One frame of the settle spring: a critically damped pull toward
    /// `target` (~150 ms to arrive), frame-rate independent.
    public static func stepSettle(_ offset: Double, target: Double, elapsedMs: Double) -> Double {
        let remaining = target - offset
        if abs(remaining) < 0.5 { return target }
        let t = 1 - exp(-elapsedMs / 55)
        return offset + remaining * t
    }
}
